src/inspector_agent.cc

```diff
@@ -0,0 +1,515 @@
+  case kInspectorHandshakeUpgrading:
+    return AcceptsConnection(socket, path);
+  case kInspectorHandshakeUpgraded:
+    agent->OnInspectorConnectionIO(socket);
+    return true;
+  case kInspectorHandshakeFailed:
+    return false;
+  default:
+    ASSERT(false);
+  }
+}
+
+// static
+void Agent::OnRemoteDataIO(uv_stream_t* stream,
+                           ssize_t read,
+                           const uv_buf_t* b) {
+  inspector_socket_t* socket =
+      reinterpret_cast<inspector_socket_t*>(stream->data);
+  Agent* agent = reinterpret_cast<Agent*>(socket->data);
+  if (read > 0) {
+    uv_mutex_lock(&agent->queue_lock_);
+    blink::protocol::String16 str(b->base, read - 1);
+    agent->message_queue_.push_back(str);
+    uv_mutex_unlock(&agent->queue_lock_);
+    free(b->base);
+
+    // TODO(pfeldman): Instead of blocking execution while debugger
+    // engages, node should wait for the run callback from the remote client
+    // and initiate its startup. This is a change to node.cc that should be
+    // upstreamed separately.
+    if (agent->wait_ &&
+        str.find(blink::protocol::String16("\"Runtime.run\"")) !=
+            std::string::npos) {
+      agent->wait_ = false;
+      uv_sem_post(&agent->start_sem_);
+    }
+
+    agent->platform_->CallOnForegroundThread(agent->parent_env_->isolate(),
+        new DispatchOnInspectorBackendTask(agent));
+    agent->parent_env_->isolate()
+        ->RequestInterrupt(InterruptCallback, agent);
+    uv_async_send(&agent->data_written_);
+  } else if (read < 0) {
+    if (agent->client_socket_ == socket) {
+      agent->client_socket_ = nullptr;
+    }
+    DisconnectAndDisposeIO(socket);
+  } else {
+    // EOF
+    if (agent->client_socket_ == socket) {
+      agent->client_socket_ = nullptr;
+      agent->platform_->CallOnForegroundThread(agent->parent_env_->isolate(),
+          new SetConnectedTask(agent, false));
+      uv_async_send(&agent->data_written_);
+    }
+  }
+  uv_cond_broadcast(&agent->pause_cond_);
+}
+
+// static
+void Agent::WriteCbIO(uv_async_t* async) {
+  auto req = reinterpret_cast<AsyncWriteRequest*>(async->data);
+  req->perform();
+  delete req;
+  uv_close(reinterpret_cast<uv_handle_t*>(async), DisposeAsyncCb);
+}
+
+void Agent::WorkerRunIO() {
+  int err;
+  sockaddr_in addr;
+  uv_tcp_t server;
+  uv_tcp_init(&child_loop_, &server);
+  uv_ip4_addr("0.0.0.0", port_, &addr);
+  server.data = this;
+  err = uv_tcp_bind(&server, (const struct sockaddr*)&addr, 0);
+  if (err == 0) {
+    err = uv_listen(reinterpret_cast<uv_stream_t*>(&server), 0,
+                    OnSocketConnectionIO);
+  }
+  if (err == 0) {
+    PrintDebuggerReadyMessage(port_);
+  } else {
+    fprintf(stderr, "Unable to open devtools socket: %s\n", uv_strerror(err));
+    assert(false);
+  }
+  if (!wait_) {
+    uv_sem_post(&start_sem_);
+  }
+  uv_run(&child_loop_, UV_RUN_DEFAULT);
+  uv_close(reinterpret_cast<uv_handle_t*>(&server), nullptr);
+  uv_run(&child_loop_, UV_RUN_NOWAIT);
```

Comment:
This should probably be UV_RUN_DEFAULT because the server handle is not guaranteed to close immediately.  (It does on UNIX platforms but Windows is more complicated.)